src/Microsoft.Health.Fhir.SqlServer/Features/ChangeFeed/SqlServerFhirResourceChangeDataStore.cs

```diff
@@ -0,0 +1,111 @@
+using Microsoft.Health.Abstractions.Data;
+using Microsoft.Health.Fhir.Core.Models;
+using Microsoft.Health.SqlServer;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.ChangeFeed
+{
+    /// <summary>
+    /// Data store for resource changes.
+    /// </summary>
+    public class SqlServerFhirResourceChangeDataStore : IChangeFeedSource<IResourceChangeData>
+    {
+        private readonly ISqlConnectionFactory _sqlConnectionFactory;
+        private readonly ILogger<SqlServerFhirResourceChangeDataStore> _logger;
+
+        // dbnetlib error value for timeout expired
+        public const short TIMEOUTEXPIRED = -2;
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="SqlServerFhirResourceChangeDataStore"/> class.
+        /// </summary>
+        /// <param name="sqlConnectionFactory">The SQL Connection factory.</param>
+        /// <param name="logger">The logger.</param>
+        public SqlServerFhirResourceChangeDataStore(ISqlConnectionFactory sqlConnectionFactory, ILogger<SqlServerFhirResourceChangeDataStore> logger)
+        {
+            EnsureArg.IsNotNull(sqlConnectionFactory, nameof(sqlConnectionFactory));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _sqlConnectionFactory = sqlConnectionFactory;
+            _logger = logger;
+        }
+
+        /// <summary>
+        ///  Returns the number of resource change records from startId.
+        /// </summary>
+        /// <param name="startId">The start id of resource change records to fetch. The start id is inclusive.</param>
+        /// <param name="pageSize">The page size for fetching resource change records.</param>
+        /// <param name="cancellationToken">Cancellation token.</param>
+        /// <returns>Resource change data rows.</returns>
+        /// <exception cref="System.ArgumentOutOfRangeException">Thrown if startId or pageSize is less than zero.</exception>
+        /// <exception cref="System.InvalidOperationException">Thrown when a method call is invalid for the object's current state.</exception>
+        /// <exception cref="Microsoft.Data.SqlClient.SqlException">Thrown when SQL Server returns a warning or error.</exception>
+        /// <exception cref="System.TimeoutException">Thrown when the time allotted for a process or operation has expired.</exception>
+        /// <exception cref="System.Exception">Thrown when errors occur during execution.</exception>
+        public async Task<IReadOnlyCollection<IResourceChangeData>> GetRecordsAsync(long startId, int pageSize, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsGte(startId, 0, nameof(startId));
+            EnsureArg.IsGte(pageSize, 0, nameof(pageSize));
+
+            var listResourceChangeData = new List<ResourceChangeData>();
+            try
+            {
+                using (SqlConnection sqlConnection = await _sqlConnectionFactory.GetSqlConnectionAsync(cancellationToken: cancellationToken))
```

Comment:
>GetSqlConnectionAsync [](http://example.com/codeflow?start=81&length=21)

How often this GetRecordsAsync would be called?

I don't know implementation details, but I'm conscern what we can create too many connections to sql and just exhaust sockets sql side.

Reply:
It would be called every second by one agent. Currently, there are no concurrent agents. The using block closes a connection automatically when it exits. Based on our test, there will only one active connection, and there will be some suspending connections that would be removed at any given moment for the publish event agent. At this time, it would be a good option that opens and closes a connection for each call, and we don't have many options to manage DB connections. We will evaluate database connection pooling options when supporting concurrent agents.

Reply:
Added a comment.